File: android/core/media/src/main/kotlin/com/dualmusic/core/media/LiveKitTokenService.kt
package com.dualmusic.core.media

import com.dualmusic.core.network.ApiClient
import com.dualmusic.core.network.Endpoint
import com.dualmusic.domain.media.LiveKitToken
import com.dualmusic.domain.media.LiveKitTokenRequest
import com.dualmusic.domain.media.MediaEndpoints
import kotlinx.serialization.json.Json

/**
 * Récupère les jetons LiveKit via `POST /livekit/token`.
 *
 * Le backend est la seule autorité qui émet des jetons signés (grants selon le rôle) :
 * l'app ne fabrique jamais de jeton localement.
 */
class LiveKitTokenService(private val api: ApiClient) {

    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    /**
     * Demande un jeton pour rejoindre une room.
     * @param roomName nom de room fourni par l'API.
     * @param isHost vrai pour l'artiste/host (droits de publication).
     */
    suspend fun token(roomName: String, isHost: Boolean = false, canPublish: Boolean = false): LiveKitToken {
        val body = json.encodeToString(
            LiveKitTokenRequest.serializer(),
            // canPublish omis (défaut false, encodeDefaults=false) → le backend applique isHost ;
            // envoyé à true pour un invité accepté (publication sans droits d'admin).
            LiveKitTokenRequest(roomName = roomName, isHost = isHost, canPublish = canPublish),
        )
        return api.request(Endpoint.post(MediaEndpoints.LIVEKIT_TOKEN, body))
    }
}

File: ios/Packages/DualMusicKit/Sources/CoreLiveMedia/LiveKitTokenService.swift
import Foundation
import CoreNetwork
import DomainModels

/// Récupère les jetons LiveKit via `POST /livekit/token`.
///
/// Le backend est la **seule autorité** qui émet des jetons signés (les grants dépendent du
/// rôle) : l'app ne fabrique jamais de jeton localement. Miroir de `LiveKitTokenService`
/// côté Android.
public struct LiveKitTokenService: Sendable {

    private let http: HTTPClient

    /// - Parameter http: client HTTP applicatif (Bearer + enveloppe gérés en amont).
    public init(http: HTTPClient) {
        self.http = http
    }

    /// Demande un jeton pour rejoindre une room.
    ///
    /// - Parameters:
    ///   - roomName: nom de room fourni par l'API (ex. `live:abc` ou `duel:123`).
    ///   - isHost: vrai pour l'artiste/host (droits de publication) ; faux pour un viewer.
    ///   - canPublish: vrai pour un invité qui publie dans SA PROPRE room (`live-guest-…`)
    ///     sans être l'hôte — le backend applique `isHost` par défaut si omis.
    /// - Returns: jeton + URL du SFU + identité du participant.
    /// - Throws: ``CoreNetwork/APIError``.
    public func token(roomName: String, isHost: Bool = false, canPublish: Bool = false) async throws -> LiveKitToken {
        try await http.request(
            .post(
                MediaEndpoints.livekitToken,
                body: LiveKitTokenRequest(roomName: roomName, isHost: isHost, canPublish: canPublish)
            ),
            as: LiveKitToken.self
        )
    }
}
